[no-unused-vars] Don’t mark `declare class` expressions as unused

<!--
If you have a problem with a specific rule, please begin your issue title with [rulename] to make it easier to search for.
I.e. "[no-unused-vars] False positive when fooing the bar"
-->

**Repro**

<!--
Include a minimal reproduction case.
Please try to avoid code that isn't directly related to the bug, as it makes it harder to investigate.
-->

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error", { "args": "none" }]
  }
}
```

```TS
declare class ResizeObserver {
  public constructor(cb: (entries: ReadonlyArray<IResizeObserverEntry>) => void)

  public disconnect(): void
  public observe(e: HTMLElement): void
}
```

**Expected Result**

No error

**Actual Result**

```
.../desktop/app/src/lib/globals.d.ts
  220:15  error  'ResizeObserver' is defined but never used  @typescript-eslint/no-unused-vars
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.0.0` |
| `@typescript-eslint/parser`        | `1.0.0` |
| `ESLint`                           | `5.8.0` |
| `node`                             | `8.12.0` |
| `npm`                              | ??? |

--- a/packages/eslint-plugin/lib/rules/no-unused-vars.js
+++ b/packages/eslint-plugin/lib/rules/no-unused-vars.js
@@ -92,6 +92,12 @@ module.exports = Object.assign({}, baseRule, {
       },
       '*[declare=true] Identifier'(node) {
         context.markVariableAsUsed(node.name);
+        const scope = context.getScope();
+        const { variableScope } = scope;
+        if (variableScope !== scope) {
+          const superVar = variableScope.set.get(node.name);
+          if (superVar) superVar.eslintUsed = true;
+        }
       }
     });
   }

--- a/packages/eslint-plugin/tests/lib/rules/no-unused-vars.js
+++ b/packages/eslint-plugin/tests/lib/rules/no-unused-vars.js
@@ -542,6 +542,13 @@ declare var Foo: {
     foo(): string
 }
     `,
+    // https://github.com/typescript-eslint/typescript-eslint/issues/106
+    `
+declare class Foo {
+    constructor(value?: any): Object;
+    foo(): string;
+}
+    `,
     `
 import foo from 'foo';
 export interface Bar extends foo.i18n {}

